### PR TITLE
Fyle custom userlist changes: Changes for adding custom emails in custom fields of type userlist

### PR DESCRIPTION
**Pull request by shishirams222**
Screenshots:

On entering search text, after debounce, email validation happens along with options will be filtered:

Screenshot 2021-04-22 at 12 04 02 AM

Only after entering a valid email, the new value can be added:
Screenshot 2021-05-06 at 10 22 46 AM

On clicking add, it will be shown at the top and will be selected:

Screenshot 2021-04-22 at 12 04 40 AM

On clicking 'Use Selected':
Screenshot 2021-04-22 at 12 05 02 AM

### CODE REVIEW

File: src/app/shared/components/fy-userlist/fy-userlist-modal/fy-userlist-modal.component.ts [hunk 1]
```diff
@@ -110,23 +114,76 @@ export class FyUserlistModalComponent implements OnInit, AfterViewInit {
...
+    if (this.allowCustomValues) {
+      this.newlyAddedItems$ = fromEvent(this.searchBarRef.nativeElement, 'keyup').pipe(
+        map((event: any) => event.srcElement.value),
+        startWith(''),
+        distinctUntilChanged(),
+        debounceTime(400),
+        switchMap((searchText) => {
+          // if newly added value is a valid email
+          var emailRegex = /^\S+@\S+\.\S{2,}$/;
+          this.invalidEmail = searchText && searchText.length > 0 && !(emailRegex.test(searchText));
```
Comment (Julias0) [inline src/app/shared/components/fy-userlist/fy-userlist-modal/fy-userlist-modal.component.ts:160]: we should do this inside a tap block

File: src/app/shared/components/fy-userlist/fy-userlist-modal/fy-userlist-modal.component.ts [hunk 2]
```diff
@@ -17,13 +16,17 @@ export class FyUserlistModalComponent implements OnInit, AfterViewInit {
...
+  @Input() allowCustomValues: boolean;
 
   value;
   eouc$: Observable<Employee[]>;
   options: { label: string, value: any, selected?: boolean }[] = [];
   selectedUsers: any[] = [];
   intialSelectedEmployees: any[] = [];
   userListCopy$: Observable<Employee[]>;
+  newlyAddedItems$: Observable<any[]>;
+  invalidEmail = false;
```
Comment (Julias0) [inline src/app/shared/components/fy-userlist/fy-userlist-modal/fy-userlist-modal.component.ts:28]: skip using any

File: src/app/shared/components/fy-userlist/fy-userlist-modal/fy-userlist-modal.component.ts [hunk 3]
```diff
@@ -17,13 +16,17 @@ export class FyUserlistModalComponent implements OnInit, AfterViewInit {
...
+  currentSelectionsCopy: any[] = [];
```
Comment (Julias0) [inline src/app/shared/components/fy-userlist/fy-userlist-modal/fy-userlist-modal.component.ts:29]: no any

File: src/app/shared/components/fy-userlist/fy-userlist-modal/fy-userlist-modal.component.ts [hunk 4]
```diff
@@ -110,23 +114,76 @@ export class FyUserlistModalComponent implements OnInit, AfterViewInit {
...
+        distinctUntilChanged(),
+        debounceTime(400),
+        switchMap((searchText) => {
+          // if newly added value is a valid email
+          var emailRegex = /^\S+@\S+\.\S{2,}$/;
+          this.invalidEmail = searchText && searchText.length > 0 && !(emailRegex.test(searchText));
+          return from(this.filteredOptions$).pipe(
+            switchMap((filteredOptions) => {
+              return this.getNewlyAddedUsers(filteredOptions).pipe(
+                map((newlyAddedItems: Partial<Employee>[] ) => {
```
Comment (Julias0) [inline src/app/shared/components/fy-userlist/fy-userlist-modal/fy-userlist-modal.component.ts:164]: exract to function

File: src/app/shared/components/fy-userlist/fy-userlist-modal/fy-userlist-modal.component.ts (hunk 2 above)
Comment (shishirams222) [reply src/app/shared/components/fy-userlist/fy-userlist-modal/fy-userlist-modal.component.ts:28]: no more any. have cleaned up

File: src/app/shared/components/fy-userlist/fy-userlist-modal/fy-userlist-modal.component.ts (hunk 1 above)
Comment (shishirams222) [reply src/app/shared/components/fy-userlist/fy-userlist-modal/fy-userlist-modal.component.ts:185]: Updated

**Review by Julias0 — APPROVED** [review]

**Pull request merged**